cli/src/test-utils/test-runner.ts

```diff
@@ -0,0 +1,300 @@
+import axios, { AxiosRequestConfig } from "axios";
+import {
+  JsonSchemaType,
+  jsonTypeSchema
+} from "../../../lib/src/generators/contract/json-schema";
+import {
+  ContractDefinition,
+  DefaultResponseDefinition,
+  EndpointDefinition,
+  TestDefinition
+} from "../../../lib/src/models/definitions";
+import { TypeNode } from "../../../lib/src/models/nodes";
+import { DataType } from "../../../lib/src/models/types";
+import { valueFromDataExpression } from "../../../lib/src/utilities/data-expression-utils";
+import { asyncForEach } from "../common/utils";
+import TestLogger from "./test-logger";
+
+export async function runTest(
+  definition: ContractDefinition,
+  stateUrl: string,
+  baseUrl: string,
+  testFilter?: TestFilter
+): Promise<boolean> {
+  let allPassed = true;
+
+  await asyncForEach(definition.endpoints, async endpoint => {
+    await asyncForEach(endpoint.tests, async test => {
+      if (testFilter) {
+        if (
+          testFilter.endpoint !== endpoint.name ||
+          (testFilter.test && testFilter.test !== test.name)
+        ) {
+          TestLogger.warn(`Test ${endpoint.name}:${test.name} skipped`);
+          return;
+        }
+      }
+      TestLogger.log(`Testing ${endpoint.name}:${test.name}`);
+      const correlatedResponse = findCorrelatedResponse(endpoint, test);
+      const result = await executeTest(
+        test,
+        stateUrl,
+        baseUrl,
+        endpoint,
+        correlatedResponse,
+        definition.types
+      );
+      if (result) {
+        TestLogger.success(`\tTest ${endpoint.name}:${test.name} passed`);
+      } else {
+        TestLogger.error(`\tTest ${endpoint.name}:${test.name} failed`);
+      }
+      allPassed = allPassed && result;
+    });
+  });
+
+  return allPassed;
+}
+
+async function executeTest(
+  test: TestDefinition,
+  stateUrl: string,
+  baseUrl: string,
+  endpoint: EndpointDefinition,
+  correlatedResponse: DefaultResponseDefinition,
+  typeStore: TypeNode[]
+): Promise<boolean> {
+  if (await executeStateSetup(test, stateUrl)) {
+    const testResult = await executeRequestUnderTest(
+      endpoint,
+      test,
+      baseUrl,
+      correlatedResponse,
+      typeStore
+    );
+    const stateTearDownResult = await executeStateTeardown(stateUrl);
+    return testResult && stateTearDownResult;
+  } else {
+    await executeStateTeardown(stateUrl);
+    return false;
+  }
+}
+
+function findCorrelatedResponse(
+  endpoint: EndpointDefinition,
+  test: TestDefinition
+): DefaultResponseDefinition {
+  const correlatedResponse =
+    endpoint.responses.find(
+      response => response.status === test.response.status
+    ) || endpoint.defaultResponse;
+  if (!correlatedResponse) {
+    throw new Error(
+      `a response with status ${
+        test.response.status
+      } was not found and a default response has not been defined`
+    );
+  }
+  return correlatedResponse;
+}
+
+function generateAxiosConfig(
```

Comment:
I'd like to see some JSDoc on each of these functions (not necessarily super extensive)